<html xmlns:mwsh="http://www.mathworks.com/namespace/mcode/v1/syntaxhighlight.dtd">
   <head>
      <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
   
      <!--
This HTML is auto-generated from an M-file.
To make changes, update the M-file and republish this document.
      -->
      <title>HIVDEMO Analyzing the origin of the Human Immunodeficiency Virus</title>
      <meta name="generator" content="MATLAB 7.0.0.3954 (R14) Prerelease">
      <meta name="date" content="2004-03-02">
      <meta name="m-file" content="hivdemo">
      <meta name="title" content="HIVDEMO Analyzing the origin of the Human Immunodeficiency Virus">
      <meta name="description" content="This demonstration shows how to construct phylogenetic trees from multiple strains of the HIV and SIV viruses."><style>
body {
  background-color: white;
}
h1 {
  color: #990000; 
  font-size: x-large;
}
h2 {
  color: #990000;
  font-size: medium;
}
p.footer {
  text-align: right;
  font-size: xx-small;
  font-weight: lighter;
  font-style: italic;
  color: gray;
}

pre.codeinput {
  margin-left: 30px;
}

span.keyword {color: #0000FF}
span.comment {color: #228B22}
span.string {color: #A020F0}
span.untermstring {color: #B20000}
span.syscmd {color: #B28C00}

pre.showbuttons {
  margin-left: 30px;
  border: solid black 2px;
  padding: 4px;
  background: #EBEFF3;
}

pre.codeoutput {
  color: gray;
  font-style: italic;
}
pre.error {
  color: red;
}
    </style></head>
   <body>
      <h1>HIVDEMO Analyzing the origin of the Human Immunodeficiency Virus</h1>
      <p>This demonstration shows how to construct phylogenetic trees from multiple strains of the HIV and SIV viruses.</p>
      <h2>Contents</h2>
      <ul>
         <li><a href="#1">Introduction</a></li>
         <li><a href="#4">Retrieve sequence information from GenBank</a></li>
         <li><a href="#8">Phylogenetic tree reconstruction</a></li>
         <li><a href="#11">Build a consensus tree</a></li>
         <li><a href="#12">Origins of the HIV virus</a></li>
      </ul>
      <h2>Introduction<a name="1"></a></h2>
      <p>Mutations accumulate in the genomes of pathogens, in this case the human/simian immunodeficiency virus, during the spread
         of an infection. This information can be used to study the history of transmission events, and also as evidence for the origins
         of the different viral strains.
      </p>
      <p>There are two characterized strains of human AIDS viruses: type 1 (HIV-1) and type 2 (HIV-2). Both strains represent cross-species
         infections. The primate reservoir of HIV-2 has been clearly identified as the sooty mangabey  (Cercocebus atys). The origin
         of HIV-1 is believed to be the common chimpanzee (Pan troglodytes).
      </p>
      <p>References:   "Origin of HIV-1 in the chimpanzee Pan troglodytes troglodytes"     Nature 397(6718), 436-41 (1999)   "Comparison
         of simian immunodeficiency virus isolates"     Nature 331(6157), 619-622 (1988)   "Genetic variability of the AIDS virus:
         nucleotide sequence analysis"     of two isolates from African patients. Cell 46 (1), 63-74 (1986)
      </p>
      <h2>Retrieve sequence information from GenBank<a name="4"></a></h2>
      <p>In this example, the variations in three longest coding regions from seventeen different isolated strains of the Human and
         Simian immunodeficiency virus are used to construct a phylogentic tree. The sequences for these virus strains can be retrieved
         from GenBank using their accession numbers. The three coding regions of interest, the gag protein, the pol polyprotein and
         the envelope polyprotein precursor, can then be extracted from the sequences using the CDS information in the GenBank records.
      </p><pre class="codeinput"><span class="comment">%        Description                   Accession  CDS:gag/pol/env</span>
data = {<span class="string">'HIV-1 (Zaire)'</span>                <span class="string">'K03454'</span>   [1 2 8]  ;
        <span class="string">'HIV1-NDK (Zaire)'</span>             <span class="string">'M27323'</span>   [1 2 8]  ;
        <span class="string">'HIV-2 (Senegal)'</span>              <span class="string">'M15390'</span>   [1 2 8]  ;
        <span class="string">'HIV2-MCN13'</span>                   <span class="string">'AY509259'</span> [1 2 8]  ;
        <span class="string">'HIV-2UC1 (IvoryCoast)'</span>        <span class="string">'L07625'</span>   [1 2 8]  ;
        <span class="string">'SIVMM251 Macaque'</span>             <span class="string">'M19499'</span>   [1 2 8]  ;
        <span class="string">'SIVAGM677A Green monkey'</span>      <span class="string">'M58410'</span>   [1 2 7]  ;
        <span class="string">'SIVlhoest L''Hoest monkeys'</span>   <span class="string">'AF075269'</span> [1 2 7]  ;
        <span class="string">'SIVcpz Chimpanzees Cameroon'</span>  <span class="string">'AF115393'</span> [1 2 8]  ;
        <span class="string">'SIVmnd5440 Mandrillus sphinx'</span> <span class="string">'AY159322'</span> [1 2 8]  ;
        <span class="string">'SIVAGM3 Green monkeys'</span>        <span class="string">'M30931'</span>   [1 2 7]  ;
        <span class="string">'SIVMM239 Simian macaque'</span>      <span class="string">'M33262'</span>   [1 2 8]  ;
        <span class="string">'CIVcpzUS Chimpanzee'</span>          <span class="string">'AF103818'</span> [1 2 8]  ;
        <span class="string">'SIVmon Cercopithecus Monkeys'</span> <span class="string">'AY340701'</span> [1 2 8]  ;
        <span class="string">'SIVcpzTAN1 Chimpanzee'</span>        <span class="string">'AF447763'</span> [1 2 8]  ;
        <span class="string">'SIVsmSL92b Sooty Mangabey'</span>    <span class="string">'AF334679'</span> [1 2 8]  ;
        };

numViruses = size(data,1)
</pre><pre class="codeoutput">
numViruses =

    16

</pre><p>You can use the <b>getgenbank</b> function to copy the data from GenBank into a structure in MATLAB. The SearchURL field of the structure contains the address
         of the actual GenBank record. You can access this record using the <b>web</b> command.
      </p><pre class="codeinput">acc_num = data{1,2}
seqs_hiv =  getgenbank(acc_num)
web(seqs_hiv(1).SearchURL)
</pre><pre class="codeoutput">
acc_num =

K03454


seqs_hiv = 

                LocusName: 'HIVELICG'
      LocusSequenceLength: '9176'
     LocusNumberofStrands: ''
            LocusTopology: 'linear'
        LocusMoleculeType: 'DNA'
     LocusGenBankDivision: 'VRL'
    LocusModificationDate: '27-FEB-2002'
               Definition: [1x66 char]
                Accession: 'K03454 X04414'
                  Version: 'K03454.1'
                       GI: '326675'
                 Keywords: []
                  Segment: []
                   Source: 'Human immunodeficiency virus 1 (HIV-1)'
           SourceOrganism: [3x59 char]
                Reference: {[1x1 struct]}
                  Comment: [1x56 char]
                 Features: [128x74 char]
                      CDS: [9x4 double]
                 Sequence: [1x9176 char]
                SearchURL: [1x105 char]
              RetrieveURL: [1x95 char]

</pre><p>Retrieve the sequence information from the NCBI GenBank database for the rest of the accession numbers.</p><pre class="codeinput"><span class="keyword">for</span> ind = 2:numViruses
    seqs_hiv(ind) = getgenbank(data{ind,2});
<span class="keyword">end</span>
</pre><p>Extract CDS for the GAG, POL, and ENV coding regions. Then extract the nucleotide sequences using the CDS pointers.</p><pre class="codeinput"><span class="keyword">for</span> ind = 1:numViruses
    temp_seq = seqs_hiv(ind).Sequence;
    temp_seq = regexprep(temp_seq,<span class="string">'[nry]'</span>,<span class="string">'a'</span>);
    CDSs = seqs_hiv(ind).CDS(data{ind,3},:);
    gag(ind).Sequence = temp_seq(CDSs(1,1):CDSs(1,2));
    pol(ind).Sequence = temp_seq(CDSs(2,1):CDSs(2,2));
    env(ind).Sequence = temp_seq(CDSs(3,1):CDSs(3,2));
<span class="keyword">end</span>
</pre><h2>Phylogenetic tree reconstruction<a name="8"></a></h2>
      <p>The <b>seqpdist</b> and <b>seqlinkage</b> commands are used to construct a phylogenetic tree for the GAG coding region using the 'Tajima-Nei' method to measure the
         distance between the sequences and the unweighted pair group method using arithmetic averages, or 'UPGMA' method, for the
         hierarchical clustering. The 'Tajima-Nei' method is only defined for nucleotides, therefore nucleotide sequences are used
         rather than the translated amino acid sequences. The distance calculation may take quite a few minutes as it is very computationally
         intensive.
      </p><pre class="codeinput">gagd = seqpdist(gag,<span class="string">'method'</span>,<span class="string">'Tajima-Nei'</span>,<span class="string">'Alphabet'</span>,<span class="string">'NT'</span>,<span class="string">'indel'</span>,<span class="string">'pair'</span>);
gagtree = seqlinkage(gagd,<span class="string">'UPGMA'</span>,data(:,1))
plot(gagtree,<span class="string">'type'</span>,<span class="string">'cladogram'</span>);
title(<span class="string">'Immunodeficieny virus (GAG protein)'</span>)
</pre><pre class="codeoutput">    Phylogenetic tree object with 16 leaves (15 branches)
</pre><img vspace="5" hspace="5" src="hivdemo_01.png"><p>Next construct a phylogenetic tree for the POL polyproteins using the 'Jukes-Cantor' method to measure distance between sequences
         and the weighted pair group method using arithmetic averages, or 'WPGMA' method, for the hierarchical clustering. The 'Jukes-Cantor'
         method is defined for amino-acids sequences, which, being significantly shorter than the corresponding nucleotide sequences,
         means that the calculation of the pairwise distances will be significantly faster.
      </p><pre class="codeinput"><span class="comment">% Convert nucleotide sequences to amino acid sequences using *nt2aa*.</span>
<span class="keyword">for</span> ind = 1:numViruses
    aagag(ind).Sequence = nt2aa(gag(ind).Sequence);
    aapol(ind).Sequence = nt2aa(pol(ind).Sequence);
    aaenv(ind).Sequence = nt2aa(env(ind).Sequence);
<span class="keyword">end</span>

<span class="comment">% Calculate the distance and linkage, and then generate the tree.</span>
pold = seqpdist(aapol,<span class="string">'method'</span>,<span class="string">'Jukes-Cantor'</span>,<span class="string">'indel'</span>,<span class="string">'pair'</span>);
poltree = seqlinkage(pold,<span class="string">'WPGMA'</span>,data(:,1))
plot(poltree,<span class="string">'type'</span>,<span class="string">'cladogram'</span>);
title(<span class="string">'Immunodeficieny virus (POL polyprotein)'</span>)
</pre><pre class="codeoutput">    Phylogenetic tree object with 16 leaves (15 branches)
</pre><img vspace="5" hspace="5" src="hivdemo_02.png"><p>Construct a phylogenetic tree for the ENV polyproteins using the normalized pairwise alignment scores as distances between
         sequences and the 'UPGMA', method for hierarchical clustering.
      </p><pre class="codeinput">envd = seqpdist(aaenv,<span class="string">'method'</span>,<span class="string">'Alignment'</span>,<span class="string">'indel'</span>,<span class="string">'score'</span>,<span class="keyword">...</span>
                    <span class="string">'ScoringMatrix'</span>,<span class="string">'Blosum62'</span>);
envtree = seqlinkage(envd,<span class="string">'UPGMA'</span>,data(:,1))
plot(envtree,<span class="string">'type'</span>,<span class="string">'cladogram'</span>);
title(<span class="string">'Immunodeficieny virus (ENV polyprotein)'</span>)
</pre><pre class="codeoutput">    Phylogenetic tree object with 16 leaves (15 branches)
</pre><img vspace="5" hspace="5" src="hivdemo_03.png"><h2>Build a consensus tree<a name="11"></a></h2>
      <p>The three trees are similar but there are some interesting differences. For example in the POL tree, the 'SIVmnd5440 Mandrillus
         sphinx' sequence is placed close to the HIV-1 strains, but in the ENV tree it is shown as being very distant to the HIV-1
         sequences. Given that the three trees show slightly different results, a consensus tree using all three regions, may give
         better general information about the complete viruses. A consensus tree can be built using a weighted average of the three
         trees.
      </p><pre class="codeinput">weights = [sum(gagd) sum(pold) sum(envd)];
weights = weights / sum(weights);
dist = gagd .* weights(1) + pold .* weights(2) + envd .* weights(3);</pre>

<p>Note that different metrics were used in the calculation of the pairwise
   distances. This could bias the consensus tree. You may wish to
   recalculate the distances for the three regions using the same metric to
   get an unbiased tree.</p>

<pre class="codeinput">tree_hiv = seqlinkage(dist,<span class="string">'average'</span>,data(:,1));
plot(tree_hiv,<span class="string">'type'</span>,<span class="string">'cladogram'</span>);
title(<span class="string">'Immunodeficieny virus (Weighted tree)'</span>)
</pre><img vspace="5" hspace="5" src="hivdemo_04.png"><h2>Origins of the HIV virus<a name="12"></a></h2>
      <p>The phylogenetic tree resulting from our analysis illustrates the presence of two clusters and some other isolated strains.
         The most compact custer includes all the HIV2 samples; at the top branch of this cluster we observe the sooty mangabey which
         has been identified as the origin of this lentivirus in humans. The cluster containing the HIV1 strain, however is not as
         compact as the HIV2 cluster. From the tree it appears that the Chimpanzee is the source of HIV1, however, the origin of the
         cross-species transmission to humans is still a matter of debate amongst HIV researchers.
      </p><pre class="codeinput"><span class="comment">% Add annotations</span>
annotation1 = annotation(gcf,<span class="string">'textarrow'</span>,[0.2875 0.3089],[0.681 0.7571],<span class="keyword">...</span>
  <span class="string">'Color'</span>,[1 0.5 0],<span class="string">'String'</span>,{<span class="string">'Possible HIV type 1 origin'</span>},<span class="keyword">...</span>
  <span class="string">'TextColor'</span>,[1 0.5 0]);

annotation(gcf,<span class="string">'textarrow'</span>,[0.4196 0.4893],[0.5929 0.5405],<span class="keyword">...</span>
  <span class="string">'Color'</span>,[1 0 0],<span class="string">'String'</span>,{<span class="string">'HIV type 2 origin'</span>},<span class="string">'TextColor'</span>,[1 0 0]);
</pre><img vspace="5" hspace="5" src="hivdemo_05.png"><p class="footer">Copyright 2003-2004 The MathWorks, Inc.<br></p>
      <!--
##### SOURCE BEGIN #####
%% HIVDEMO Analyzing the origin of the Human Immunodeficiency Virus
% This demonstration shows how to construct phylogenetic trees from
% multiple strains of the HIV and SIV viruses.

%   Copyright 2003-2004 The MathWorks, Inc.
%   $Revision: 1.1.2.2 $  $Date: 2004/03/01 23:09:43 $



%% Introduction
% Mutations accumulate in the genomes of pathogens, in this case the
% human/simian immunodeficiency virus, during the spread of an infection.
% This information can be used to study the history of transmission events,
% and also as evidence for the origins of the different viral strains. 

%%
% There are two characterized strains of human AIDS viruses: type 1 (HIV-1)
% and type 2 (HIV-2). Both strains represent cross-species infections. The
% primate reservoir of HIV-2 has been clearly identified as the sooty
% mangabey  (Cercocebus atys). The origin of HIV-1 is believed to be the
% common chimpanzee (Pan troglodytes).

%%
% References:
%   "Origin of HIV-1 in the chimpanzee Pan troglodytes troglodytes" 
%     Nature 397(6718), 436-41 (1999)
%   "Comparison of simian immunodeficiency virus isolates"
%     Nature 331(6157), 619-622 (1988)
%   "Genetic variability of the AIDS virus: nucleotide sequence analysis"
%     of two isolates from African patients. Cell 46 (1), 63-74 (1986)

%% Retrieve sequence information from GenBank
% In this example, the variations in three longest coding regions from
% seventeen different isolated strains of the Human and Simian
% immunodeficiency virus are used to construct a phylogentic tree. The
% sequences for these virus strains can be retrieved from GenBank using
% their accession numbers. The three coding regions of interest, the gag
% protein, the pol polyprotein and the envelope polyprotein precursor, can
% then be extracted from the sequences using the CDS information in the
% GenBank records.

%        Description                   Accession  CDS:gag/pol/env 
data = {'HIV-1 (Zaire)'                'K03454'   [1 2 8]  ;
        'HIV1-NDK (Zaire)'             'M27323'   [1 2 8]  ;
        'HIV-2 (Senegal)'              'M15390'   [1 2 8]  ;
        'HIV2-MCN13'                   'AY509259' [1 2 8]  ;
        'HIV-2UC1 (IvoryCoast)'        'L07625'   [1 2 8]  ;
        'SIVMM251 Macaque'             'M19499'   [1 2 8]  ;
        'SIVAGM677A Green monkey'      'M58410'   [1 2 7]  ;
        'SIVlhoest L''Hoest monkeys'   'AF075269' [1 2 7]  ;
        'SIVcpz Chimpanzees Cameroon'  'AF115393' [1 2 8]  ;
        'SIVmnd5440 Mandrillus sphinx' 'AY159322' [1 2 8]  ;
        'SIVAGM3 Green monkeys'        'M30931'   [1 2 7]  ;
        'SIVMM239 Simian macaque'      'M33262'   [1 2 8]  ;
        'CIVcpzUS Chimpanzee'          'AF103818' [1 2 8]  ;
        'SIVmon Cercopithecus Monkeys' 'AY340701' [1 2 8]  ;
        'SIVcpzTAN1 Chimpanzee'        'AF447763' [1 2 8]  ;
        'SIVsmSL92b Sooty Mangabey'    'AF334679' [1 2 8]  ;
        };

numViruses = size(data,1)
    
%%
% You can use the *getgenbank* function to copy the data from GenBank into
% a structure in MATLAB. The SearchURL field of the structure contains the
% address of the actual GenBank record. You can access this record using
% the *web* command.

acc_num = data{1,2}
seqs_hiv =  getgenbank(acc_num)
web(seqs_hiv(1).SearchURL)

%%
% Retrieve the sequence information from the NCBI GenBank database for the
% rest of the accession numbers. 
for ind = 2:numViruses
    seqs_hiv(ind) = getgenbank(data{ind,2});
end

%%
% Extract CDS for the GAG, POL, and ENV coding regions. Then extract the
% nucleotide sequences using the CDS pointers.
for ind = 1:numViruses
    temp_seq = seqs_hiv(ind).Sequence; 
    temp_seq = regexprep(temp_seq,'[nry]','a');
    CDSs = seqs_hiv(ind).CDS(data{ind,3},:);
    gag(ind).Sequence = temp_seq(CDSs(1,1):CDSs(1,2));
    pol(ind).Sequence = temp_seq(CDSs(2,1):CDSs(2,2));
    env(ind).Sequence = temp_seq(CDSs(3,1):CDSs(3,2));
end

%% Phylogenetic tree reconstruction
% The *seqpdist* and *seqlinkage* commands are used to construct a
% phylogenetic tree for the GAG coding region using the 'Tajima-Nei' method
% to measure the distance between the sequences and the unweighted pair
% group method using arithmetic averages, or 'UPGMA' method, for the
% hierarchical clustering. The 'Tajima-Nei' method is only defined for
% nucleotides, therefore nucleotide sequences are used rather than the
% translated amino acid sequences. The distance calculation may take quite
% a few minutes as it is very computationally intensive.

gagd = seqpdist(gag,'method','Tajima-Nei','Alphabet','NT','indel','pair');
gagtree = seqlinkage(gagd,'UPGMA',data(:,1))
plot(gagtree,'type','cladogram');putwithebackgroundinlastfigure
title('Immunodeficieny virus (GAG protein)')

%%
% Next construct a phylogenetic tree for the POL polyproteins using the
% 'Jukes-Cantor' method to measure distance between sequences and the
% weighted pair group method using arithmetic averages, or 'WPGMA' method,
% for the hierarchical clustering. The 'Jukes-Cantor' method is defined for
% amino-acids sequences, which, being significantly shorter than the
% corresponding nucleotide sequences, means that the calculation of the
% pairwise distances will be significantly faster.

% Convert nucleotide sequences to amino acid sequences using *nt2aa*.
for ind = 1:numViruses
    aagag(ind).Sequence = nt2aa(gag(ind).Sequence);
    aapol(ind).Sequence = nt2aa(pol(ind).Sequence);
    aaenv(ind).Sequence = nt2aa(env(ind).Sequence);
end

% Calculate the distance and linkage, and then generate the tree.
pold = seqpdist(aapol,'method','Jukes-Cantor','indel','pair');
poltree = seqlinkage(pold,'WPGMA',data(:,1))
plot(poltree,'type','cladogram');putwithebackgroundinlastfigure
title('Immunodeficieny virus (POL polyprotein)')

%%
% Construct a phylogenetic tree for the ENV polyproteins using the
% normalized pairwise alignment scores as distances between sequences and
% the 'UPGMA', method for hierarchical clustering. 

envd = seqpdist(aaenv,'method','Alignment','indel','score',...
                    'ScoringMatrix','Blosum62');
envtree = seqlinkage(envd,'UPGMA',data(:,1))
plot(envtree,'type','cladogram');putwithebackgroundinlastfigure
title('Immunodeficieny virus (ENV polyprotein)')

%% Build a consensus tree
% The three trees are similar but there are some interesting differences.
% For example in the POL tree, the 'SIVmnd5440 Mandrillus sphinx' sequence
% is placed close to the HIV-1 strains, but in the ENV tree it is shown as
% being very distant to the HIV-1 sequences.
% Given that the three trees show slightly different results, a consensus
% tree using all three regions, may give better general information about
% the complete viruses. A consensus tree can be built using a weighted
% average of the three trees. 

weights = [sum(gagd) sum(pold) sum(envd)];
weights = weights / sum(weights);
dist = gagd .* weights(1) + pold .* weights(2) + envd .* weights(3);

% Note that different metrics were used in the calculation of the pairwise
% distances. This could bias the consensus tree. You may wish to
% recalculate the distances for the three regions using the same metric to
% get an unbiased tree.

tree_hiv = seqlinkage(dist,'average',data(:,1));
plot(tree_hiv,'type','cladogram');putwithebackgroundinlastfigure
title('Immunodeficieny virus (Weighted tree)')

%% Origins of the HIV virus
% The phylogenetic tree resulting from our analysis illustrates the presence
% of two clusters and some other isolated strains. The most compact custer
% includes all the HIV2 samples; at the top branch of this cluster 
% we observe the sooty mangabey which has been identified as the
% origin of this lentivirus in humans. The cluster containing the HIV1
% strain, however is not as compact as the HIV2 cluster. From the tree it
% appears that the Chimpanzee is the source of HIV1, however, the origin of
% the cross-species transmission to humans is still a matter of debate
% amongst HIV researchers. 

% Add annotations
annotation1 = annotation(gcf,'textarrow',[0.2875 0.3089],[0.681 0.7571],...
  'Color',[1 0.5 0],'String',{'Possible HIV type 1 origin'},...
  'TextColor',[1 0.5 0]);

annotation(gcf,'textarrow',[0.4196 0.4893],[0.5929 0.5405],...
  'Color',[1 0 0],'String',{'HIV type 2 origin'},'TextColor',[1 0 0]);
 
 

##### SOURCE END #####
-->
   </body>
</html>