<html xmlns:mwsh="http://www.mathworks.com/namespace/mcode/v1/syntaxhighlight.dtd">
   <head>
      <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
   
      <!--
This HTML is auto-generated from an M-file.
To make changes, update the M-file and republish this document.
      -->
      <title>PRIMATESDEMO Building a phylogenetic tree for the hominidae species.</title>
      <meta name="generator" content="MATLAB 7.0.0.3450 (R14) Beta 2">
      <meta name="date" content="2004-01-23">
      <meta name="m-file" content="primatesdemo">
      <meta name="title" content="PRIMATESDEMO Building a phylogenetic tree for the hominidae species.">
      <meta name="description" content="A phylogenetic tree is constructed from mtDNA sequences for the Hominidae taxa (also known as pongidae). This family embraces the gorillas, chimpanzees, orangutans and the humans."><style>
body {
  background-color: white;
}
h1 {
  color: #990000; 
  font-size: x-large;
}
h2 {
  color: #990000;
  font-size: medium;
}
p.footer {
  text-align: right;
  font-size: xx-small;
  font-weight: lighter;
  font-style: italic;
  color: gray;
}

pre.codeinput {
  margin-left: 30px;
}

span.keyword {color: #0000FF}
span.comment {color: #228B22}
span.string {color: #A020F0}
span.untermstring {color: #B20000}
span.syscmd {color: #B28C00}

pre.showbuttons {
  margin-left: 30px;
  border: solid black 2px;
  padding: 4px;
  background: #EBEFF3;
}

pre.codeoutput {
  color: gray;
  font-style: italic;
}
pre.error {
  color: red;
}
    </style></head>
   <body>
      <h1>PRIMATESDEMO Building a phylogenetic tree for the hominidae species.</h1>
      <p>A phylogenetic tree is constructed from mtDNA sequences for the Hominidae taxa (also known as pongidae). This family embraces
         the gorillas, chimpanzees, orangutans and the humans.
      </p>
      <h2>Contents</h2>
      <ul>
         <li><a href="#1">Introduction</a></li>
         <li><a href="#3">Phylogenetic tree for five species</a></li>
         <li><a href="#6">Scaling up the phylogenetic tree</a></li>
         <li><a href="#9">Exploring the phylogenetic tree</a></li>
      </ul>
      <h2>Introduction<a name="1"></a></h2>
      <p>The mitochondrial D-loop is one of the fastest mutating sequence regions in animal DNA. Therefore, useful for comparing closely
         related organisms. The origin of modern man is a highly debated issue that has recently been tackled by using mtDNA sequences.
         The limited genetic variability of human mtDNA has been explained in terms of a recent common genetic ancestry, thus implying
         that all modern-population mtDNAs originated from a single woman who lived in Africa less than 200,000 years.
      </p>
      <p>References:   "Molecular analysis of Neanderthal DNA from the northern Caucasus"   Nature 404 (6777), 490-493 (2000),   "Genes
         and languages in Europe: an analysis of mitochondrial lineages"   Genome Res. 5 (1), 42-52 (1995),   "Neandertal DNA sequences
         and the origin of modern humans"   Cell 90 (1), 19-30 (1997),   "Mitochondrial DNA variation and biogeography of eastern gorillas"
           Mol. Ecol. 10 (9), 2241-2247 (2001)
      </p>
      <h2>Phylogenetic tree for five species<a name="3"></a></h2>
      <p>Get some data from GenBank, these are the accession codes for mitochondrial D-loop sequences that have been isolated for different
         hominidae species.
      </p><pre class="codeinput"><span class="comment">%          Description            Accession ID</span>
data = {<span class="string">'German_Neanderthal'</span>      <span class="string">'AF011222'</span>;
        <span class="string">'Russian_Neanderthal'</span>     <span class="string">'AF254446'</span>;
        <span class="string">'European_Human'</span>          <span class="string">'X90314'</span>  ;
        <span class="string">'Mountain_Gorilla_Rwanda'</span> <span class="string">'AF089820'</span>;
        <span class="string">'Chimp_Troglodytes'</span>       <span class="string">'AF176766'</span>;
       };

<span class="keyword">for</span> ind = 1:5
    seqs(ind).Header   = data{ind,1};
    seqs(ind).Sequence = getgenbank(data{ind,2},<span class="string">'sequenceonly'</span>,<span class="string">'true'</span>);
<span class="keyword">end</span>
</pre><p>Compute pairwise distances using the 'Jukes-Cantor' formula and the phylogenetic tree with the 'UPGMA' distance method. Since
         the sequences are not pre-aligned, <b>seqpdist</b> will pairwise align them before computing the distances.
      </p><pre class="codeinput">distances = seqpdist(seqs,<span class="string">'Method'</span>,<span class="string">'Jukes-Cantor'</span>,<span class="string">'Alpha'</span>,<span class="string">'DNA'</span>);
tree = seqlinkage(distances,<span class="string">'UPGMA'</span>,seqs)
</pre><pre class="codeoutput">    Phylogenetic tree object with 5 leaves (4 branches)
</pre><p>Render the phylogenetic tree</p><pre class="codeinput">h = plot(tree,<span class="string">'orient'</span>,<span class="string">'bottom'</span>);
ylabel(<span class="string">'Evolutionary distance'</span>)
set(h.terminalNodeLabels,<span class="string">'Rotation'</span>,-45)
</pre><img vspace="5" hspace="5" src="primatesdemo_01.png"><h2>Scaling up the phylogenetic tree<a name="6"></a></h2>
      <p>Scale our experiment up by including some more mtDNA sequences. We also get them from GeneBank.</p><pre class="codeinput"><span class="comment">%          Description            Accession ID</span>
data2 = {<span class="string">'Puti_Orangutan'</span>          <span class="string">'AF451972'</span>;
         <span class="string">'Jari_Orangutan'</span>          <span class="string">'AF451964'</span>;
         <span class="string">'Western_Lowland_Gorilla'</span> <span class="string">'AY079510'</span>;
         <span class="string">'Eastern_Lowland_Gorilla'</span> <span class="string">'AF050738'</span>;
         <span class="string">'Chimp_Schweinfurthii'</span>    <span class="string">'AF176722'</span>;
         <span class="string">'Chimp_Vellerosus'</span>        <span class="string">'AF315498'</span>;
         <span class="string">'Chimp_Verus'</span>             <span class="string">'AF176731'</span>;
       };

<span class="keyword">for</span> ind = 1:7
    seqs(ind+5).Header   = data2{ind,1};
    seqs(ind+5).Sequence = getgenbank(data2{ind,2},<span class="string">'sequenceonly'</span>,<span class="string">'true'</span>);
<span class="keyword">end</span>
</pre><p>Compute pairwise distances and the hierarchical linkage.</p><pre class="codeinput">distances = seqpdist(seqs,<span class="string">'Method'</span>,<span class="string">'Jukes-Cantor'</span>,<span class="string">'Alpha'</span>,<span class="string">'DNA'</span>);
tree = seqlinkage(distances,<span class="string">'UPGMA'</span>,seqs);
</pre><p>Render the phylogenetic tree</p><pre class="codeinput">h = plot(tree,<span class="string">'orient'</span>,<span class="string">'bottom'</span>);
ylabel(<span class="string">'Evolutionary distance'</span>)
set(h.terminalNodeLabels,<span class="string">'Rotation'</span>,-45)
</pre><img vspace="5" hspace="5" src="primatesdemo_02.png"><h2>Exploring the phylogenetic tree<a name="9"></a></h2>
      <p>Find the closet species to the 'European Human' entry (3).</p><pre class="codeinput">names = get(tree,<span class="string">'LeafNames'</span>)
[h_all,h_leaves] = select(tree,<span class="string">'reference'</span>,3,<span class="string">'criteria'</span>,<span class="string">'distance'</span>,<span class="string">'threshold'</span>,0.6);
<span class="comment">% h_all has now all the nodes within 0.6 of patristic distance to the 'European</span>
<span class="comment">% Human' leave.</span>
<span class="comment">% h_leaves has only the leaf nodes within 0.6 of patristic distance to (3)</span>
subtree_names = names(h_leaves)
</pre><pre class="codeoutput">
names = 

    'German_Neanderthal'
    'Russian_Neanderthal'
    'European_Human'
    'Chimp_Troglodytes'
    'Chimp_Schweinfurthii'
    'Chimp_Verus'
    'Chimp_Vellerosus'
    'Puti_Orangutan'
    'Jari_Orangutan'
    'Mountain_Gorilla_Rwanda'
    'Eastern_Lowland_Gorilla'
    'Western_Lowland_Gorilla'


subtree_names = 

    'German_Neanderthal'
    'Russian_Neanderthal'
    'European_Human'
    'Chimp_Troglodytes'
    'Chimp_Schweinfurthii'
    'Chimp_Verus'

</pre><p>Reduce the tree to the sub-branch of interest</p><pre class="codeinput">leaves_to_prune = ~h_leaves;
pruned_tree = prune(tree,leaves_to_prune)
h = plot(pruned_tree,<span class="string">'orient'</span>,<span class="string">'bottom'</span>);
ylabel(<span class="string">'Evolutionary distance'</span>)
set(h.terminalNodeLabels,<span class="string">'Rotation'</span>,-30)
</pre><pre class="codeoutput">    Phylogenetic tree object with 6 leaves (5 branches)
</pre><img vspace="5" hspace="5" src="primatesdemo_03.png"><p>With <b>view</b> you can further explore/edit the phylogenetic tree using an interactive tool. See also <b>phytreetool</b>.
      </p><pre class="codeinput">view(tree,h_leaves)
</pre><img vspace="5" hspace="5" src="primatesdemo_04.png"><p class="footer">Copyright 2003-2004 The MathWorks, Inc.<br></p>
      <!--
##### SOURCE BEGIN #####
%% PRIMATESDEMO Building a phylogenetic tree for the hominidae species.
% A phylogenetic tree is constructed from mtDNA sequences for the Hominidae
% taxa (also known as pongidae). This family embraces the gorillas,
% chimpanzees, orangutans and the humans.

%   Copyright 2003-2004 The MathWorks, Inc.
%   $Revision:  $  $Date: $


%% Introduction
% The mitochondrial D-loop is one of the fastest mutating sequence regions
% in animal DNA. Therefore, useful for comparing closely related organisms.
% The origin of modern man is a highly debated issue that has recently been 
% tackled by using mtDNA sequences. The limited genetic variability of
% human mtDNA has been explained in terms of a recent common genetic
% ancestry, thus implying that all modern-population mtDNAs originated from
% a single woman who lived in Africa less than 200,000 years.

%%
% References:
%   "Molecular analysis of Neanderthal DNA from the northern Caucasus"
%   Nature 404 (6777), 490-493 (2000),
%   "Genes and languages in Europe: an analysis of mitochondrial lineages"
%   Genome Res. 5 (1), 42-52 (1995),
%   "Neandertal DNA sequences and the origin of modern humans"
%   Cell 90 (1), 19-30 (1997),
%   "Mitochondrial DNA variation and biogeography of eastern gorillas"
%   Mol. Ecol. 10 (9), 2241-2247 (2001)

%% Phylogenetic tree for five species
% Get some data from GenBank, these are the accession codes for
% mitochondrial D-loop sequences that have been isolated for different
% hominidae species.  

%          Description            Accession ID
data = {'German_Neanderthal'      'AF011222';
        'Russian_Neanderthal'     'AF254446';
        'European_Human'          'X90314'  ;
        'Mountain_Gorilla_Rwanda' 'AF089820';
        'Chimp_Troglodytes'       'AF176766';
       };
   
for ind = 1:5       
    seqs(ind).Header   = data{ind,1};
    seqs(ind).Sequence = getgenbank(data{ind,2},'sequenceonly','true');
end

%%
% Compute pairwise distances using the 'Jukes-Cantor' formula and the
% phylogenetic tree with the 'UPGMA' distance method. Since the sequences
% are not pre-aligned, *seqpdist* will pairwise align them before computing
% the distances. 
distances = seqpdist(seqs,'Method','Jukes-Cantor','Alpha','DNA');
tree = seqlinkage(distances,'UPGMA',seqs)

%%
% Render the phylogenetic tree 
h = plot(tree,'orient','bottom');putwithebackgroundinlastfigure;set(h.axes,'Position',[.1 .25 .75 .7])
ylabel('Evolutionary distance')
set(h.terminalNodeLabels,'Rotation',-45)

%% Scaling up the phylogenetic tree
% Scale our experiment up by including some more mtDNA sequences. We also
% get them from GeneBank. 

%          Description            Accession ID
data2 = {'Puti_Orangutan'          'AF451972';
         'Jari_Orangutan'          'AF451964';
         'Western_Lowland_Gorilla' 'AY079510';
         'Eastern_Lowland_Gorilla' 'AF050738';
         'Chimp_Schweinfurthii'    'AF176722';
         'Chimp_Vellerosus'        'AF315498';
         'Chimp_Verus'             'AF176731';
       };

for ind = 1:7       
    seqs(ind+5).Header   = data2{ind,1};
    seqs(ind+5).Sequence = getgenbank(data2{ind,2},'sequenceonly','true');
end

%%
% Compute pairwise distances and the hierarchical linkage.
distances = seqpdist(seqs,'Method','Jukes-Cantor','Alpha','DNA');
tree = seqlinkage(distances,'UPGMA',seqs);

%%
% Render the phylogenetic tree
h = plot(tree,'orient','bottom');putwithebackgroundinlastfigure;set(h.axes,'Position',[.1 .25 .75 .7])
ylabel('Evolutionary distance')
set(h.terminalNodeLabels,'Rotation',-45)

%% Exploring the phylogenetic tree
% Find the closet species to the 'European Human' entry (3).
names = get(tree,'LeafNames')
[h_all,h_leaves] = select(tree,'reference',3,'criteria','distance','threshold',0.6);
% h_all has now all the nodes within 0.6 of patristic distance to the 'European
% Human' leave.
% h_leaves has only the leaf nodes within 0.6 of patristic distance to (3)
subtree_names = names(h_leaves)

%%
% Reduce the tree to the sub-branch of interest
leaves_to_prune = ~h_leaves;
pruned_tree = prune(tree,leaves_to_prune)
h = plot(pruned_tree,'orient','bottom');putwithebackgroundinlastfigure
ylabel('Evolutionary distance')
set(h.terminalNodeLabels,'Rotation',-30)

%%
% With *view* you can further explore/edit the phylogenetic tree using an
% interactive tool. See also *phytreetool*.
view(tree,h_leaves)


##### SOURCE END #####
-->
   </body>
</html>